controllers/hvpa_controller.go

```diff
@@ -164,43 +164,70 @@ func (r *HvpaReconciler) getSelectorFromHvpa(instance *autoscalingv1alpha1.Hvpa)
 	if err != nil {
 		log.Error(err, "Error getting target using targetRef.", "Will skip", instance.Name)
-		return nil, err
+		return nil, nil, err
 	}
 
-	selectorMap, found, err := unstructured.NestedMap(target.Object, "spec", "selector")
-	if err != nil {
+	var selector labels.Selector
+
+	// TODO use the Scale subresource if possible
+	if selectorMap, found, err := unstructured.NestedMap(target.Object, "spec", "selector"); err != nil {
 		log.Error(err, "Not able to get the selectorMap from target.", "Will skip", instance.Name)
-		return nil, err
-	}
-	if !found {
+		return nil, nil, err
+	} else if !found {
 		log.V(2).Info("Target doesn't have selector", "will skip HVPA", instance.Name)
-		return nil, err
-	}
+		return nil, nil, err
+	} else {
+		labelSelector := &metav1.LabelSelector{}
+		selectorStr, err := json.Marshal(selectorMap)
+		err = json.Unmarshal(selectorStr, &labelSelector)
+		if err != nil {
+			log.Error(err, "Error in reading selector string.", "will skip", instance.Name)
+			return nil, nil, err
+		}
 
-	labelSelector := &metav1.LabelSelector{}
-	selectorStr, err := json.Marshal(selectorMap)
-	err = json.Unmarshal(selectorStr, &labelSelector)
-	if err != nil {
-		log.Error(err, "Error in reading selector string.", "will skip", instance.Name)
-		return nil, err
+		selector, err = metav1.LabelSelectorAsSelector(labelSelector)
+		if err != nil {
+			log.Error(err, "Error in getting label selector", "will skip", instance.Name)
+			return nil, nil, err
+		}
 	}
 
-	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
-	if err != nil {
-		log.Error(err, "Error in getting label selector", "will skip", instance.Name)
-		return nil, err
+	// TODO use the Scale subresource if possible
+	if replicas, found, err := unstructured.NestedInt64(target.Object, "status", "replicas"); err != nil {
+		log.Error(err, "Not able to get the status replicas from target.", "Will skip", instance.Name)
+		return nil, nil, err
+	} else if !found {
+		log.V(2).Info("Target doesn't have status replicas", "will skip HVPA", instance.Name)
+		return nil, nil, err
+	} else {
+		statusReplicas := int32(replicas)
+		return selector, &statusReplicas, nil
 	}
-	return selector, nil
 }
 
-// manageCache manages the global map of HVPAs
-func (r *HvpaReconciler) manageCache(instance *autoscalingv1alpha1.Hvpa) {
-	selector, err := r.getSelectorFromHvpa(instance)
+// reconcileScaleStatus reconciles the status fields relevant to the Scale subresource.
+func (r *HvpaReconciler) reconcileScaleStatus(instance *autoscalingv1alpha1.Hvpa) error {
+	selector, statusReplicas, err := r.getScaleStatusFromTarget(instance)
 	if err != nil {
-		log.Error(err, "Error in getting label selector", "will skip", instance.Name)
-		return
+		return err
+	}
+
+	selectorStr := selector.String()
+	if oldSelector := instance.Status.TargetSelector; oldSelector == nil || *oldSelector != selectorStr {
+		instance.Status.TargetSelector = &selectorStr
```

Comment:
Should we DeepCopy instance before updating? Currently `Get` already returns a DeepCopy of cached value, however, this could change in future.
Refer: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/internal/cache_reader.go#L74

Also, all the `status` calculations happens at [one place](https://github.com/gardener/hvpa-controller/blob/b7976d0cd47bd7438f08e3954c5d0f9d2cb60b06/controllers/hvpa_controller.go#L1164-L1204) currently. Would it be safer to continue doing that. This would make it easy to maintain code.

Reply:
👍 Done